File: scripts/import_cutoffs.py
"""
Import cutoff data from CSV into the SQLite database.

Usage:
    python scripts/import_cutoffs.py <csv_file> <GATE|JEE>

Example:
    python scripts/import_cutoffs.py data/josaa_2024.csv JEE
    python scripts/import_cutoffs.py data/ccmt_2024.csv GATE

CSV must have these columns (case-insensitive):
    institute, program, category, year, round, openRank (or open_rank),
    closeRank (or close_rank), state, instituteType (or institute_type)

Category values must be: GEN, OBC, SC, ST, EWS, GEN-PwD, OBC-PwD, SC-PwD, ST-PwD
(JoSAA uses OPEN/OBC-NCL — these are mapped automatically)
"""

import csv
import sqlite3
import sys
import os
import uuid

DB_PATH = os.path.join(os.path.dirname(__file__), '..', 'dev.db')

# Map JoSAA/CCMT category names to our DB values
CATEGORY_MAP = {
    'OPEN': 'GEN',
    'OBC-NCL': 'OBC',
    'SC': 'SC',
    'ST': 'ST',
    'EWS': 'EWS',
    'OPEN-PWD': 'GEN-PwD',
    'OBC-NCL-PWD': 'OBC-PwD',
    'SC-PWD': 'SC-PwD',
    'ST-PWD': 'ST-PwD',
    'GEN': 'GEN',
    'OBC': 'OBC',
    'GEN-PWD': 'GEN-PwD',
    'OBC-PWD': 'OBC-PwD',
}


def normalize_col(headers: list[str]) -> dict[str, str]:
    """Build a mapping from normalized header names to actual CSV column names."""
    mapping = {}
    alias = {
        'institute': ['institute', 'college', 'institution'],
        'program': ['program', 'course', 'branch', 'programme'],
        'category': ['category', 'quota', 'seat_type', 'seattype'],
        'year': ['year'],
        'round': ['round', 'round_no', 'roundno'],
        'openRank': ['openrank', 'open_rank', 'opening_rank', 'openingrank'],
        'closeRank': ['closerank', 'close_rank', 'closing_rank', 'closingrank'],
        'state': ['state', 'institute_state'],
        'instituteType': ['institutetype', 'institute_type', 'type', 'inst_type'],
    }
    lower_headers = {h.lower().replace(' ', '_'): h for h in headers}
    for field, candidates in alias.items():
        for c in candidates:
            if c in lower_headers:
                mapping[field] = lower_headers[c]
                break
    return mapping


def import_csv(csv_file: str, exam_type: str):
    if not os.path.exists(csv_file):
        print(f"Error: file not found: {csv_file}")
        sys.exit(1)

    if not os.path.exists(DB_PATH):
        print(f"Error: database not found at {DB_PATH}")
        print("Run `npx prisma db push` first to create the database.")
        sys.exit(1)

    conn = sqlite3.connect(DB_PATH)
    cursor = conn.cursor()
    inserted = skipped = 0

    with open(csv_file, 'r', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        col = normalize_col(reader.fieldnames or [])

        missing = [k for k in ('institute', 'program', 'openRank', 'closeRank') if k not in col]
        if missing:
            print(f"Error: CSV is missing required columns: {missing}")
            print(f"Found columns: {reader.fieldnames}")
            sys.exit(1)

        for row in reader:
            try:
                raw_cat = row.get(col.get('category', ''), 'GEN').strip().upper()
                category = CATEGORY_MAP.get(raw_cat, raw_cat)

                cursor.execute("""
                    INSERT OR IGNORE INTO Cutoff
                    (id, examType, year, institute, program, category, round,
                     openRank, closeRank, state, instituteType, createdAt, updatedAt)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, datetime('now'), datetime('now'))
                """, (
                    str(uuid.uuid4()).replace('-', ''),
                    exam_type,
                    int(row.get(col.get('year', ''), 2024) or 2024),
                    row[col['institute']].strip(),
                    row[col['program']].strip(),
                    category,
                    int(row.get(col.get('round', ''), 1) or 1),
                    int(str(row[col['openRank']]).replace(',', '') or 0),
                    int(str(row[col['closeRank']]).replace(',', '') or 0),
                    row.get(col.get('state', ''), '').strip() or None,
                    row.get(col.get('instituteType', ''), '').strip() or None,
                ))
                inserted += 1
            except Exception as e:
                print(f"  Skipped row: {e}")
                skipped += 1

    conn.commit()
    conn.close()
    print(f"Done. Inserted: {inserted}, Skipped: {skipped}")
    print(f"Database: {DB_PATH}")


if __name__ == '__main__':
    if len(sys.argv) < 3:
        print("Usage: python import_cutoffs.py <csv_file> <GATE|JEE>")
        print("Example: python import_cutoffs.py data/josaa_2024.csv JEE")
        sys.exit(1)

    import_csv(sys.argv[1], sys.argv[2].upper())

File: scripts/import-cutoffs.mjs
import { PrismaClient } from '@prisma/client'
import { PrismaPg } from '@prisma/adapter-pg'
import { Pool } from 'pg'
import fs from 'fs'
import path from 'path'

const ROOT = path.resolve(path.dirname(new URL(import.meta.url).pathname), '..')
const DATA = path.join(ROOT, 'data')
const YEAR = 2025
const BATCH_SIZE = 1000

function getPoolConfig(url) {
  let needsSsl = url.includes('supabase.com')

  try {
    const parsed = new URL(url)
    needsSsl = needsSsl || parsed.searchParams.get('sslmode') === 'require'
  } catch {
    // Let pg surface the connection-string error with its normal message.
  }

  return {
    connectionString: url,
    connectionTimeoutMillis: 10000,
    ssl: needsSsl ? { rejectUnauthorized: false } : undefined,
  }
}

const CATEGORY_MAP = {
  OPEN: 'GEN',
  'OPEN-PWD': 'GEN-PwD',
  'OPEN-(PWD)': 'GEN-PwD',
  'OBC-NCL': 'OBC',
  'OBC-NCL-PWD': 'OBC-PwD',
  'OBC-NCL-(PWD)': 'OBC-PwD',
  'SC-PWD': 'SC-PwD',
  'SC-(PWD)': 'SC-PwD',
  'ST-PWD': 'ST-PwD',
  'ST-(PWD)': 'ST-PwD',
  'EWS-PWD': 'EWS-PwD',
  'EWS-(PWD)': 'EWS-PwD',
  GEN: 'GEN',
  OBC: 'OBC',
  SC: 'SC',
  ST: 'ST',
  EWS: 'EWS',
  'GEN-PWD': 'GEN-PwD',
  'OBC-PWD': 'OBC-PwD',
  'GEN-PwD': 'GEN-PwD',
  'OBC-PwD': 'OBC-PwD',
  'SC-PwD': 'SC-PwD',
  'ST-PwD': 'ST-PwD',
  'EWS-PwD': 'EWS-PwD',
}

const NIT_STATE = {
  'national institute of technology karnataka': 'Karnataka',
  'national institute of technology warangal': 'Telangana',
  'national institute of technology tiruchirappalli': 'Tamil Nadu',
  'national institute of technology calicut': 'Kerala',
  'national institute of technology rourkela': 'Odisha',
  'national institute of technology jamshedpur': 'Jharkhand',
  'national institute of technology patna': 'Bihar',
  'national institute of technology silchar': 'Assam',
  'national institute of technology agartala': 'Tripura',
  'national institute of technology meghalaya': 'Meghalaya',
  'national institute of technology manipur': 'Manipur',
  'national institute of technology mizoram': 'Mizoram',
  'national institute of technology nagaland': 'Nagaland',
  'national institute of technology srinagar': 'J&K',
  'national institute of technology hamirpur': 'Himachal Pradesh',
  'national institute of technology kurukshetra': 'Haryana',
  'national institute of technology delhi': 'Delhi',
  'national institute of technology durgapur': 'West Bengal',
  'national institute of technology raipur': 'Chhattisgarh',
  'national institute of technology, andhra': 'Andhra Pradesh',
  'national institute of technology arunachal': 'Arunachal Pradesh',
  'national institute of technology uttarakhand': 'Uttarakhand',
  'national institute of technology goa': 'Goa',
  'national institute of technology puducherry': 'Puducherry',
  'national institute of technology sikkim': 'Sikkim',
  'maulana azad national institute': 'Madhya Pradesh',
  'motilal nehru national institute': 'Uttar Pradesh',
  'sardar vallabhbhai national institute': 'Gujarat',
  'visvesvaraya national institute': 'Maharashtra',
  'dr. b r ambedkar national institute': 'Punjab',
  'malaviya national institute': 'Rajasthan',
}

function parseCsv(text) {
  const rows = []
  let row = []
  let field = ''
  let inQuotes = false

  for (let i = 0; i < text.length; i += 1) {
    const ch = text[i]
    const next = text[i + 1]

    if (ch === '"') {
      if (inQuotes && next === '"') {
        field += '"'
        i += 1
      } else {
        inQuotes = !inQuotes
      }
      continue
    }

    if (ch === ',' && !inQuotes) {
      row.push(field)
      field = ''
      continue
    }

    if ((ch === '\n' || ch === '\r') && !inQuotes) {
      if (ch === '\r' && next === '\n') i += 1
      row.push(field)
      if (row.some(value => value !== '')) rows.push(row)
      row = []
      field = ''
      continue
    }

    field += ch
  }

  row.push(field)
  if (row.some(value => value !== '')) rows.push(row)
  const [headers, ...data] = rows
  return data.map(values => Object.fromEntries(headers.map((header, index) => [
    header.replace(/^\uFEFF/, '').trim(),
    (values[index] ?? '').trim(),
  ])))
}

function readCsv(file) {
  return parseCsv(fs.readFileSync(path.join(DATA, file), 'utf8'))
}

function intValue(value, fallback = 0) {
  const parsed = Number.parseInt(String(value ?? '').replace(/,/g, ''), 10)
  return Number.isFinite(parsed) ? parsed : fallback
}

function normalizeCategory(value) {
  const raw = String(value || 'GEN').trim()
  const key = raw.toUpperCase().replace(/\s+/g, '-')
  return CATEGORY_MAP[key] || CATEGORY_MAP[raw] || raw
}

function norm(value) {
  return String(value).toLowerCase().replace(/[^a-z0-9 ]/g, '')
}

function guessState(institute, given) {
  if (given) return given
  const low = norm(institute)
  for (const [needle, state] of Object.entries(NIT_STATE)) {
    if (low.includes(norm(needle))) return state
  }
  return null
}

function jeeRows(file, source) {
  let skipped = 0
  const rows = readCsv(file).flatMap(row => {
    const institute = row.institute?.trim()
    const openRank = intValue(row.openRank)
    const closeRank = intValue(row.closeRank)
    if (!institute || (openRank === 0 && closeRank === 0)) {
      skipped += 1
      return []
    }
    return [{
      examType: 'JEE',
      year: intValue(row.year, YEAR),
      institute,
      program: row.program?.trim() || '',
      category: normalizeCategory(row.category),
      round: intValue(row.round, 1),
      openRank,
      closeRank,
      openScore: 0,
      closeScore: 0,
      paper: null,
      gender: row.gender?.trim() || null,
      quota: row.quota?.trim() || null,
      source,
      state: guessState(institute, row.state?.trim()),
      instituteType: row.instituteType?.trim() || 'GFTI',
    }]
  })
  return { rows, skipped }
}

function ccmtRows(file) {
  let skipped = 0
  const rows = readCsv(file).flatMap(row => {
    const institute = row.institute?.trim()
    const openScore = intValue(row.openScore)
    const closeScore = intValue(row.closeScore)
    if (!institute || (openScore === 0 && closeScore === 0)) {
      skipped += 1
      return []
    }
    return [{
      examType: 'GATE',
      year: intValue(row.year, YEAR),
      institute,
      program: row.program?.trim() || '',
      category: normalizeCategory(row.category),
      round: intValue(row.round, 0),
      openRank: 0,
      closeRank: 0,
      openScore,
      closeScore,
      paper: row.paper?.trim() || null,
      gender: null,
      quota: null,
      source: file.startsWith('gate_') ? 'COAP' : 'CCMT',
      state: row.state?.trim() || null,
      instituteType: row.instituteType?.trim() || (institute.startsWith('IIT') ? 'IIT' : null),
    }]
  })
  return { rows, skipped }
}

async function createMany(prisma, rows) {
  let inserted = 0
  for (let i = 0; i < rows.length; i += BATCH_SIZE) {
    const batch = rows.slice(i, i + BATCH_SIZE)
    const result = await prisma.cutoff.createMany({ data: batch })
    inserted += result.count
    process.stdout.write(`\rInserted ${inserted.toLocaleString()} / ${rows.length.toLocaleString()}`)
  }
  process.stdout.write('\n')
  return inserted
}

async function main() {
  if (!process.env.DATABASE_URL) {
    throw new Error('DATABASE_URL is required')
  }

  const pool = new Pool(getPoolConfig(process.env.DATABASE_URL))
  const prisma = new PrismaClient({ adapter: new PrismaPg(pool) })

  try {
    console.log('Reading CSV files...')
    const inputs = [
      jeeRows('josaa_2025_cutoffs.csv', 'JOSAA'),
      jeeRows('csab_2025_cutoffs.csv', 'CSAB'),
      ccmtRows('ccmt_2025_cutoffs.csv'),
      ccmtRows('gate_2025_cutoffs.csv'),
    ]
    const rows = inputs.flatMap(input => input.rows)
    const skipped = inputs.reduce((sum, input) => sum + input.skipped, 0)

    console.log(`Prepared ${rows.length.toLocaleString()} rows. Skipped ${skipped.toLocaleString()} invalid rows.`)
    console.log(`Deleting existing ${YEAR} JEE/GATE rows...`)
    const deleted = await prisma.cutoff.deleteMany({
      where: { year: YEAR, examType: { in: ['JEE', 'GATE'] } },
    })
    console.log(`Deleted ${deleted.count.toLocaleString()} rows.`)

    const inserted = await createMany(prisma, rows)
    const [jee, gate, total] = await Promise.all([
      prisma.cutoff.count({ where: { examType: 'JEE', year: YEAR } }),
      prisma.cutoff.count({ where: { examType: 'GATE', year: YEAR } }),
      prisma.cutoff.count(),
    ])

    console.log(`Done. Inserted ${inserted.toLocaleString()} rows.`)
    console.log(`Counts: JEE=${jee.toLocaleString()} GATE=${gate.toLocaleString()} Total=${total.toLocaleString()}`)
  } finally {
    await prisma.$disconnect()
    await pool.end()
  }
}

main().catch(error => {
  console.error(error)
  process.exit(1)
})
